torch/export/exported_program.py

```diff
@@ -307,6 +263,49 @@ def _decompose_and_get_gm_with_new_signature_constants(
     )
     from torch.fx.experimental.symbolic_shapes import ShapeEnv
 
+    # Note [Seperating decomp_table into CIA decomps and non-CIA decomps]
+    # At this point, we have a decomp_table that contains decomp behaviour for
+    # both CIA and post-autograd ops.
+    # We need to separate the op into two categories:
+    # 1. CIA op: These are the ops that we want to override
+    #    CompositeImplicitAutograd decomp for. For them, we need to use _override_composite_implicit_decomp
+    #    context manager to plumb it through AOTDispatcher
+    # 2. Non-CIA op: These ops are only relevant after AOTDIspatcher runs, so just
+    #    checking if they are statically functional is enough.
+
+    all_preservable_cia_ops = _collect_all_valid_cia_ops()
+    core_aten_decomp = core_aten_decompositions()
+    cia_ops_to_callable = {}
+    for op in list(decomp_table.keys()):
+        # In the user specified decomp table, we want to assert that it is something
+        # we can customize.
+        # TODO We should fix the core aten decomp table to filter out mutating/aliasing ops
+        # One example is: aten.addcdiv.out
+        if op not in core_aten_decomp:
+            _assert_valid_to_preserve(op)
+        # if it is a valid CIA op we can mess with in export, we check if it is:
+        #  1. Has been marked as to be decomposed. Example:
+        #        decomp_table = decomp_table_to_core_aten()
+        #        del decomp_table[aten.linear]
+        #     In this case, user says decompose everything except for aten.linear
+        #  2. Has been marked with custom decomp behavour. Example:
+        #        decomp_table = {aten.linear: some_op}
+        # For (1), we want to remove all the CIA ops that weren't handled by user as
+        # it suggests they are safe to decompose, so we should remove from preservable_list.
+        # for (2), we just plumb the custom decomp to AOTDIspatcher.
+        # In both cases, we want to remove this CIA op from the decomp_table as it is special
+        # handled.
+        if op in all_preservable_cia_ops:
+            if decomp_table[op] is _is_special_op_to_decompose:
+                all_preservable_cia_ops.remove(op)
+            else:
+                cia_ops_to_callable[op] = decomp_table[op]
+                all_preservable_cia_ops.remove(op)
+            del decomp_table[op]
```

Comment:
just a nit, but there's probably no reason to delete the op from the decomp table too?

I guess we should expect that if the op is CIA and we are using its decomp, then we expect the CIA decomp to run before we ever hit the decomp table. Although doesn't hurt to keep it in the decomp table for safety.

Reply:
Yep technically we don't need to. But i just wanted to have a clear seperation between python decomp and export decomp. The former is the rest of dispatching system expects.